### PR TITLE
Refactoring proxy handling to be done in strategies

### PR DESCRIPTION
**Pull request by AndrewBastin**
This PR intends to move the proxy handling code to the individual strategies along with that, this PR introduces Postwoman request proxy support for Firefox Extension powered requests.

*NOTE: To test the Firefox extension code, you will have to load the extension locally. The extension that can be downloaded from the Firefox Add-on's page does not hook into localhost pages for security reasons, for more info on how to do that, visit [https://github.com/AndrewBastin/postwoman-firefox](https://github.com/AndrewBastin/postwoman-firefox)*

### CODE REVIEW

**Comment by AndrewBastin** [comment]
Also didn't notice till now.... Ticket No. 500 !!!!! wooo :rofl:

File: functions/strategies/AxiosStrategy.js [hunk 1]
```diff
@@ -1,8 +1,23 @@
...
+
+const axiosWithoutProxy = async (req, _store) => {
   const res = await axios(req);
   return res;
 };
 
+const axiosStrategy = (req, store) => {
+  if (store.state.postwoman.settings.PROXY_ENABLED)
+    return axiosWithProxy(req, store);
+  else return axiosWithoutProxy(req, store);
```
Comment (jamesgeorge007) [inline functions/strategies/AxiosStrategy.js]:
```js
  if (store.state.postwoman.settings.PROXY_ENABLED) {
    return axiosWithProxy(req, store);
  }
  return axiosWithoutProxy(req, store);
```

File: functions/strategies/FirefoxStrategy.js [hunk 2]
```diff
@@ -1,5 +1,22 @@
-const firefoxStrategy = (req, _store) => new Promise((resolve, reject) => {
 
+const firefoxWithProxy = (req, store) => new Promise((resolve, reject) => {
+  const eventListener = (event) => {
+    window.removeEventListener("firefoxExtSendRequestComplete", event);
+
+    if (event.detail.error) reject(JSON.parse(event.detail.error));
+    else resolve(JSON.parse(event.detail.response));
```
Comment (jamesgeorge007) [inline functions/strategies/FirefoxStrategy.js]:
```js
    if (event.detail.error) {
      reject(JSON.parse(event.detail.error));
    } else {
      resolve(JSON.parse(event.detail.response));
     }
```
```

File: functions/strategies/FirefoxStrategy.js [hunk 3]
```diff
@@ -12,4 +29,10 @@ const firefoxStrategy = (req, _store) => new Promise((resolve, reject) => {
   window.firefoxExtSendRequest(req);
 });
 
+const firefoxStrategy = (req, store) => {
+  if (store.state.postwoman.settings.PROXY_ENABLED)
+    return firefoxWithProxy(req, store);
+  else return firefoxWithoutProxy(req, store);
```
Comment (jamesgeorge007) [inline functions/strategies/FirefoxStrategy.js]:
```js
  if (store.state.postwoman.settings.PROXY_ENABLED) {
    return firefoxWithProxy(req, store);
  }
  return firefoxWithoutProxy(req, store);
```

**Comment by AndrewBastin** [comment]
@jamesgeorge007 check latest commits

File: functions/strategies/AxiosStrategy.js [hunk 4]
```diff
@@ -1,8 +1,24 @@
...
+const axiosWithoutProxy = async (req, _store) => {
   const res = await axios(req);
   return res;
 };
 
+const axiosStrategy = (req, store) => {
+  if (store.state.postwoman.settings.PROXY_ENABLED) {
+    return axiosWithProxy(req, store);
+  }
+  else return axiosWithoutProxy(req, store);
```
Comment (jamesgeorge007) [inline functions/strategies/AxiosStrategy.js]:
```suggestion
  return axiosWithoutProxy(req, store);
```

File: functions/strategies/FirefoxStrategy.js (hunk 2 above)
Comment (jamesgeorge007) [reply functions/strategies/FirefoxStrategy.js]: You haven't made this change yet!

File: functions/strategies/FirefoxStrategy.js [hunk 5]
```diff
@@ -12,4 +29,11 @@ const firefoxStrategy = (req, _store) => new Promise((resolve, reject) => {
   window.firefoxExtSendRequest(req);
 });
 
+const firefoxStrategy = (req, store) => {
+  if (store.state.postwoman.settings.PROXY_ENABLED) {
+    return firefoxWithProxy(req, store);
+  }
+  else return firefoxWithoutProxy(req, store);
```
Comment (jamesgeorge007) [inline functions/strategies/FirefoxStrategy.js]:
```suggestion
  return firefoxWithoutProxy(req, store);
```

File: functions/strategies/FirefoxStrategy.js (hunk 2 above)
Comment (AndrewBastin) [reply functions/strategies/FirefoxStrategy.js]: It's such a tiny detail and it is like 4AM here, soooooo, sleep deprivation just kicked in sorry :stuck_out_tongue:

**Comment by AndrewBastin** [comment]
So sorry for the mess, I gotta get some sleep :sleepy:

**Pull request merged**